File: spring-web-java/src/main/java/spring/web/java/config/security/JwtExceptionFilter.java
package spring.web.java.config.security;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.UnsupportedJwtException;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.ServletRequest;
import jakarta.servlet.ServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.http.MediaType;
import org.springframework.web.filter.GenericFilterBean;
import spring.web.java.constant.ApiResponseCode;
import spring.web.java.dto.response.ApiResponse;

import java.io.IOException;


@RequiredArgsConstructor
public class JwtExceptionFilter extends GenericFilterBean {

    private final ObjectMapper objectMapper;

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
        try {
            chain.doFilter(request, response);
        } catch (UnsupportedJwtException e) {
            setResponse(response, ApiResponseCode.UNSUPPORTED_TOKEN);
        } catch (ExpiredJwtException e) {
            setResponse(response, ApiResponseCode.EXPIRED_TOKEN);
        } catch (JwtException e) {
            setResponse(response, ApiResponseCode.BAD_TOKEN);
        }
    }

    private void setResponse(ServletResponse response, ApiResponseCode apiResponseCode) throws RuntimeException, IOException {
        String result = objectMapper.writeValueAsString(new ApiResponse<>(apiResponseCode));

        response.setContentType(MediaType.APPLICATION_JSON_VALUE);
        response.setCharacterEncoding("UTF-8");
        response.getWriter().write(result);
    }
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/filter/JwtExceptionFilter.kt
package spring.web.kotlin.filter

import com.fasterxml.jackson.databind.ObjectMapper
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.UnsupportedJwtException
import jakarta.servlet.FilterChain
import jakarta.servlet.ServletRequest
import jakarta.servlet.ServletResponse
import org.springframework.http.MediaType
import org.springframework.web.filter.GenericFilterBean
import spring.web.kotlin.constant.ApiResponseCode
import spring.web.kotlin.dto.response.ApiResponse

class JwtExceptionFilter(
    private val objectMapper: ObjectMapper
) : GenericFilterBean() {
    override fun doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain) {
        runCatching { chain.doFilter(request, response) }
            .onFailure { e ->
                when (e) {
                    is UnsupportedJwtException -> setResponse(response, ApiResponseCode.UNSUPPORTED_TOKEN)
                    is ExpiredJwtException -> setResponse(response, ApiResponseCode.EXPIRED_TOKEN)
                    is JwtException -> setResponse(response, ApiResponseCode.BAD_TOKEN)
                }
            }
    }

    private fun setResponse(response: ServletResponse, apiResponseCode: ApiResponseCode) {
        val result = objectMapper.writeValueAsString(ApiResponse<ApiResponseCode>(apiResponseCode))

        response.contentType = MediaType.APPLICATION_JSON_VALUE
        response.characterEncoding = "UTF-8"
        response.writer.write(result)
    }
}